ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -83,21 +83,96 @@ type RayClusterReconciler struct {
 // +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=get;list;watch;create;delete
 // Reconcile used to bridge the desired state with the current state
 func (r *RayClusterReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
-	_ = r.Log.WithValues("raycluster", request.NamespacedName)
-	log.Info("reconciling RayCluster", "cluster name", request.Name)
+	var err error
+
+	// Try to fetch the Event instance
+	event := &v1.Event{}
+	if err = r.Get(context.TODO(), request.NamespacedName, event); err == nil {
+		return r.eventReconcile(request, event)
+	}
 
-	// Fetch the RayCluster instance
+	// Try to fetch the RayCluster instance
 	instance := &rayiov1alpha1.RayCluster{}
-	if err := r.Get(context.TODO(), request.NamespacedName, instance); err != nil {
-		if errors.IsNotFound(err) {
-			log.Info("Read request instance not found error!")
-		} else {
-			log.Error(err, "Read request instance error!")
+	if err = r.Get(context.TODO(), request.NamespacedName, instance); err == nil {
+		return r.rayClusterReconcile(request, instance)
+	}
+
+	// No match found
+	if errors.IsNotFound(err) {
+		log.Info("Read request instance not found error!", "name", request.NamespacedName)
+	} else {
+		log.Error(err, "Read request instance error!")
+	}
+	// Error reading the object - requeue the request.
+	return ctrl.Result{}, client.IgnoreNotFound(err)
+}
+
+func (r *RayClusterReconciler) eventReconcile(request ctrl.Request, event *v1.Event) (ctrl.Result, error) {
+	var unhealthyPod *corev1.Pod
+	pods := corev1.PodList{}
+
+	// we only care about pod events
+	if event.InvolvedObject.Kind != "Pod" || event.Type != "Warning" || event.Reason != "Unhealthy" ||
+		!strings.Contains(event.Message, "Readiness probe failed") {
+		return ctrl.Result{}, nil
+	}
+
+	_ = r.Log.WithValues("event", request.NamespacedName)
+	log.Info("reconcile RayCluster Event", "event name", request.Name)
+
+	if err := r.List(context.TODO(), &pods,
+		client.InNamespace(event.InvolvedObject.Namespace)); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, item := range pods.Items {
+		if item.Name == event.InvolvedObject.Name {
+			if unhealthyPod != nil {
+				return ctrl.Result{}, fmt.Errorf("duplicate pod found")
+			}
+			unhealthyPod = &item
+		}
+	}
+
+	if unhealthyPod == nil || unhealthyPod.Annotations == nil {
+		log.Info("pod not found or no valid annotations", "pod name", event.InvolvedObject.Name)
+		return ctrl.Result{}, nil
+	}
+
+	if enabledString, ok := unhealthyPod.Annotations[common.RayHAEnabledAnnotationKey]; ok {
+		if strings.ToLower(enabledString) != "true" {
+			log.Info("HA not enabled skipping event reconcile for pod.", "pod name", unhealthyPod.Name)
```

Comment:
Should we increase the required verbosity of this log to avoid log spam when HA is not enabled? log.V(10) ?

Reply:
Oh but we've somehow detected the pod is unhealthy which (a) requires the probe (b) is uncommon.
Never mind, don't mind me.